cdap-runtime-ext-dataproc/src/main/java/io/cdap/cdap/runtime/spi/common/ArtifactCacheManager.java

```diff
@@ -0,0 +1,182 @@
+        Map<String, Integer> artifactCount;
+        if (blob == null || !blob.exists()) {
+          LOG.debug("Creating ArtifactsCacheCounter {} at {}", CACHED_ARTIFACTS_USAGE_COUNT, filePath);
+          artifactCount = new HashMap<>();
+          BlobInfo createBlob = BlobInfo.newBuilder(blobId).setContentType(CONTENT_TYPE_JSON).build();
+          try {
+            client.create(createBlob, GSON.toJson(artifactCount).getBytes(StandardCharsets.UTF_8),
+                          Storage.BlobTargetOption.doesNotExist());
+          } catch (StorageException e) {
+            if (e.getCode() != HttpURLConnection.HTTP_PRECON_FAILED) {
+              throw e;
+            }
+          }
+          blob = client.get(blobId);
+        }
+        artifactCount = GSON.fromJson(new String(blob.getContent(), StandardCharsets.UTF_8),
+                                      new TypeToken<Map<String, Integer>>() {
+                                      }.getType());
+        for (String file : files) {
+          int count = artifactCount.getOrDefault(file, 0) + changeValue;
+          if (count <= 0) {
+            if (count < 0) {
+              LOG.error("Cache usage count less than 0 for {} in {}", file, getPath(cachePath,
+                                                                                    CACHED_ARTIFACTS_USAGE_COUNT));
+            }
+            artifactCount.remove(file);
+            setCustomTimeOnArtifactAndReleaseHold(client, bucket, file, cachePath);
+          } else {
+            artifactCount.put(file, count);
+          }
+        }
+        try {
+          WritableByteChannel writer = blob.writer(Storage.BlobWriteOption.generationMatch());
+          writer.write(ByteBuffer.wrap(GSON.toJson(artifactCount).getBytes(StandardCharsets.UTF_8)));
+          writer.close();
+        } catch (StorageException | IOException e) {
+          LOG.error("Exception while writing to artifacts cache counter file", e);
+          throw e;
+        }
+        break;
+      } catch (Exception e) {
+        LOG.error("Exception while updating artifacts cache counter, retrying operation.", e);
+      }
+    }
+  }
+
+  private static void setCustomTimeOnArtifactAndReleaseHold(Storage client, String bucket, String file, String path) {
+    BlobInfo blobInfo = BlobInfo.newBuilder(bucket, getPath(path, file))
+      .setCustomTime(System.currentTimeMillis())
+      .setTemporaryHold(false).build();
+    client.update(blobInfo);
+  }
+
+  private static void storeCachedArtifactsForRun(Storage client, String bucket, Set<String> files, String filePath) {
+    BlobInfo createBlob = BlobInfo.newBuilder(bucket, filePath).setContentType(CONTENT_TYPE_JSON).build();
+    try {
+      client.create(createBlob, GSON.toJson(files).getBytes(StandardCharsets.UTF_8),
+                    Storage.BlobTargetOption.doesNotExist());
+    } catch (StorageException e) {
+      if (e.getCode() != HttpURLConnection.HTTP_PRECON_FAILED) {
+        throw e;
```

Comment:
Wondering if this will ever happen? We probably can't much about it here though.